asserts/constraint_test.go

```diff
@@ -0,0 +1,535 @@
+	c.Assert(err, IsNil)
+
+	err = domatch(vals(`
+foo: FOO
+bar:
+  bar1: BAR1
+  bar2: BAR2
+  bar3: BAR3
+baz: BAZ
+`), nil)
+	c.Check(err, IsNil)
+
+	err = domatch(vals(`
+foo: FOO
+bar: BAZ
+baz: BAZ
+`), nil)
+	c.Check(err, ErrorMatches, `attribute "bar" must be a map`)
+
+	err = domatch(vals(`
+foo: FOO
+bar:
+  bar1: BAR1
+  bar2: BAR22
+  bar3: BAR3
+baz: BAZ
+`), nil)
+	c.Check(err, ErrorMatches, `attribute "bar\.bar2" value "BAR22" does not match \^\(BAR2\)\$`)
+
+	err = domatch(vals(`
+foo: FOO
+bar:
+  bar1: BAR1
+  bar2:
+    bar22: true
+  bar3: BAR3
+baz: BAZ
+`), nil)
+	c.Check(err, ErrorMatches, `attribute "bar\.bar2" must be a scalar or list`)
+}
+
+func (s *attrMatcherSuite) TestAlternative(c *C) {
+	m, err := asserts.ParseHeaders([]byte(`attrs:
+  -
+    foo: FOO
+    bar: BAR
+  -
+    foo: FOO
+    bar: BAZ`))
+	c.Assert(err, IsNil)
+
+	domatch, err := asserts.CompileAttrMatcher(m["attrs"])
+	c.Assert(err, IsNil)
+
+	values := map[string]interface{}{
+		"foo": "FOO",
+		"bar": "BAR",
+		"baz": "BAZ",
+	}
+	err = domatch(values, nil)
+	c.Check(err, IsNil)
+
+	values = map[string]interface{}{
+		"foo": "FOO",
+		"bar": "BAZ",
+		"baz": "BAZ",
+	}
+	err = domatch(values, nil)
+	c.Check(err, IsNil)
+
+	values = map[string]interface{}{
+		"foo": "FOO",
+		"bar": "BARR",
+		"baz": "BAR",
+	}
+	err = domatch(values, nil)
+	c.Check(err, ErrorMatches, `no alternative matches: attribute "bar" value "BARR" does not match \^\(BAR\)\$`)
+}
+
+func (s *attrMatcherSuite) TestNestedAlternative(c *C) {
+	m, err := asserts.ParseHeaders([]byte(`attrs:
+  foo: FOO
+  bar:
+    bar1: BAR1
+    bar2:
+      - BAR2
+      - BAR22`))
+	c.Assert(err, IsNil)
+
+	domatch, err := asserts.CompileAttrMatcher(m["attrs"].(map[string]interface{}))
+	c.Assert(err, IsNil)
+
+	err = domatch(vals(`
+foo: FOO
+bar:
+  bar1: BAR1
+  bar2: BAR2
+`), nil)
+	c.Check(err, IsNil)
+
+	err = domatch(vals(`
+foo: FOO
+bar:
+  bar1: BAR1
+  bar2: BAR22
+`), nil)
+	c.Check(err, IsNil)
+
+	err = domatch(vals(`
+foo: FOO
+bar:
+  bar1: BAR1
+  bar2: BAR3
+`), nil)
+	c.Check(err, ErrorMatches, `no alternative for attribute "bar\.bar2" matches: attribute "bar\.bar2" value "BAR3" does not match \^\(BAR2\)\$`)
+}
+
+func (s *attrMatcherSuite) TestAlternativeMatchingStringList(c *C) {
+	toMatch := vals(`
+write:
+ - /var/tmp
+ - /var/lib/snapd/snapshots
+`)
+	m, err := asserts.ParseHeaders([]byte(`attrs:
+  write: /var/(tmp|lib/snapd/snapshots)`))
+	c.Assert(err, IsNil)
+
+	domatch, err := asserts.CompileAttrMatcher(m["attrs"].(map[string]interface{}))
+	c.Assert(err, IsNil)
+
+	err = domatch(toMatch, nil)
+	c.Check(err, IsNil)
+
+	m, err = asserts.ParseHeaders([]byte(`attrs:
+  write:
+    - /var/tmp
+    - /var/lib/snapd/snapshots`))
+	c.Assert(err, IsNil)
+
+	domatchLst, err := asserts.CompileAttrMatcher(m["attrs"].(map[string]interface{}))
+	c.Assert(err, IsNil)
+
+	err = domatchLst(toMatch, nil)
+	c.Check(err, IsNil)
```

Comment:
Would it make sense to test the error case here too or is this already tested enough in the other ones?